rust/lance-encoding/src/encodings/physical/bitpack_fastlanes.rs

```diff
@@ -0,0 +1,1547 @@
+                });
+                Ok(EncodedArray {
+                    data: encoded,
+                    encoding,
+                })
+            }
+            _ => {
+                Err(Error::InvalidInput {
+                    source: "Bitpacking only supports fixed width data blocks or a nullable data block with fixed width data block inside or a all null data block".into(),
+                    location: location!(),
+                })
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct BitpackedForNonNegScheduler {
+    compressed_bit_width: u64,
+    uncompressed_bits_per_value: u64,
+    buffer_offset: u64,
+}
+
+fn locate_chunk_start(scheduler: &BitpackedForNonNegScheduler, relative_row_num: u64) -> u64 {
+    let chunk_size = ELEMS_PER_CHUNK * scheduler.compressed_bit_width / 8;
+    relative_row_num / ELEMS_PER_CHUNK * chunk_size
+}
+
+fn locate_chunk_end(scheduler: &BitpackedForNonNegScheduler, relative_row_num: u64) -> u64 {
+    let chunk_size = ELEMS_PER_CHUNK * scheduler.compressed_bit_width / 8;
+    relative_row_num / ELEMS_PER_CHUNK * chunk_size + chunk_size
+}
+
+impl BitpackedForNonNegScheduler {
+    pub fn new(
+        compressed_bit_width: u64,
+        uncompressed_bits_per_value: u64,
+        buffer_offset: u64,
+    ) -> Self {
+        Self {
+            compressed_bit_width,
+            uncompressed_bits_per_value,
+            buffer_offset,
+        }
+    }
+}
+
+impl PageScheduler for BitpackedForNonNegScheduler {
+    fn schedule_ranges(
+        &self,
+        ranges: &[std::ops::Range<u64>],
+        scheduler: &Arc<dyn crate::EncodingsIo>,
+        top_level_row: u64,
+    ) -> BoxFuture<'static, Result<Box<dyn PrimitivePageDecoder>>> {
+        // can we get here?
+        if ranges.is_empty() {
+            panic!("cannot schedule empty ranges");
+        }
+        let mut byte_ranges = vec![];
+
+        // map one bytes to multiple ranges, one bytes has at least one range corresponding to it
+        let mut bytes_idx_to_range_indices = vec![];
+        let first_byte_range = std::ops::Range {
+            start: self.buffer_offset + locate_chunk_start(self, ranges[0].start),
+            end: self.buffer_offset + locate_chunk_end(self, ranges[0].end - 1),
+        }; // the ranges are half-open
+        byte_ranges.push(first_byte_range);
+        bytes_idx_to_range_indices.push(vec![ranges[0].clone()]);
+
+        for (i, range) in ranges.iter().enumerate().skip(1) {
+            let this_start = locate_chunk_start(self, range.start);
+            let this_end = locate_chunk_end(self, range.end - 1);
+
+            // when the current range start is in the same chunk as the previous range's end, we colaesce this two bytes ranges
+            // when the current range start is not in the same chunk as the previous range's end, we create a new bytes range
+            if this_start == locate_chunk_start(self, ranges[i - 1].end - 1) {
```

Comment:
Instead of re-running `locate_chunk_start` can we look at the last item in `byte_ranges`?

Reply:
actually, the last item in `byte_ranges` is calculated using `locate_chunk_start(range.start)` and `locate_chunk_end(range.end - 1)` and here I am comparing `this_start` with `locate_chunk_start(range.end - 1) to see if I want to coalesce `bytes_ranges`